File: code/Api/controllers/userController.js
const pool = require('../config/database');
const { hashPassword } = require('../utils/password');
const logger = require('../utils/logger');

class UserController {
  async getUserList(req, res) {
    try {
      const { page = 1, pageSize = 10, username, status } = req.query;
      const offset = (page - 1) * pageSize;

      let whereClause = 'WHERE 1=1';
      const params = [];

      if (username) {
        whereClause += ' AND username LIKE ?';
        params.push(`%${username}%`);
      }

      if (status !== undefined) {
        whereClause += ' AND status = ?';
        params.push(status);
      }

      const [users] = await pool.query(`
        SELECT id, username, email, phone, real_name, status, last_login_time, last_login_ip, created_at, updated_at
        FROM users
        ${whereClause}
        ORDER BY created_at DESC
        LIMIT ? OFFSET ?
      `, [...params, parseInt(pageSize), offset]);

      const [countResult] = await pool.query(`
        SELECT COUNT(*) as total
        FROM users
        ${whereClause}
      `, params);

      const total = countResult[0].total;

      for (const user of users) {
        const [roles] = await pool.query(`
          SELECT r.id, r.role_name, r.role_code
          FROM roles r
          INNER JOIN user_roles ur ON r.id = ur.role_id
          WHERE ur.user_id = ?
        `, [user.id]);
        user.roles = roles;
      }

      res.json({
        success: true,
        data: {
          list: users,
          total,
          page: parseInt(page),
          pageSize: parseInt(pageSize)
        }
      });
    } catch (error) {
      logger.error('获取用户列表失败:', error);
      res.status(500).json({
        success: false,
        message: '获取用户列表失败'
      });
    }
  }

  async getUserById(req, res) {
    try {
      const { id } = req.params;

      const [users] = await pool.query(
        'SELECT id, username, email, phone, real_name, status, last_login_time, last_login_ip, created_at, updated_at FROM users WHERE id = ?',
        [id]
      );

      if (users.length === 0) {
        return res.status(404).json({
          success: false,
          message: '用户不存在'
        });
      }

      const [roles] = await pool.query(`
        SELECT r.id, r.role_name, r.role_code
        FROM roles r
        INNER JOIN user_roles ur ON r.id = ur.role_id
        WHERE ur.user_id = ?
      `, [id]);

      users[0].roles = roles;

      res.json({
        success: true,
        data: users[0]
      });
    } catch (error) {
      logger.error('获取用户详情失败:', error);
      res.status(500).json({
        success: false,
        message: '获取用户详情失败'
      });
    }
  }

  async createUser(req, res) {
    try {
      const { username, password, email, phone, realName, roleIds, status = 1 } = req.body;

      const [existingUsers] = await pool.query(
        'SELECT id FROM users WHERE username = ? OR email = ?',
        [username, email]
      );

      if (existingUsers.length > 0) {
        return res.status(400).json({
          success: false,
          message: '用户名或邮箱已存在'
        });
      }

      const hashedPassword = await hashPassword(password);

      const [result] = await pool.query(
        'INSERT INTO users (username, password, email, phone, real_name, status, created_by) VALUES (?, ?, ?, ?, ?, ?, ?)',
        [username, hashedPassword, email, phone, realName, status, req.user.id]
      );

      if (roleIds && roleIds.length > 0) {
        const roleValues = roleIds.map(roleId => [result.insertId, roleId]);
        await pool.query(
          'INSERT INTO user_roles (user_id, role_id) VALUES ?',
          [roleValues]
        );
      }

      logger.info(`创建用户成功: ${username}, 操作人: ${req.user.username}`);

      res.status(201).json({
        success: true,
        message: '创建用户成功',
        data: {
          id: result.insertId,
          username
        }
      });
    } catch (error) {
      logger.error('创建用户失败:', error);
      res.status(500).json({
        success: false,
        message: '创建用户失败'
      });
    }
  }

  async updateUser(req, res) {
    try {
      const { id } = req.params;
      const { email, phone, realName, roleIds, status } = req.body;

      const [users] = await pool.query('SELECT id FROM users WHERE id = ?', [id]);

      if (users.length === 0) {
        return res.status(404).json({
          success: false,
          message: '用户不存在'
        });
      }

      await pool.query(
        'UPDATE users SET email = ?, phone = ?, real_name = ?, status = ?, updated_by = ? WHERE id = ?',
        [email, phone, realName, status, req.user.id, id]
      );

      if (roleIds !== undefined) {
        await pool.query('DELETE FROM user_roles WHERE user_id = ?', [id]);

        if (roleIds.length > 0) {
          const roleValues = roleIds.map(roleId => [id, roleId]);
          await pool.query(
            'INSERT INTO user_roles (user_id, role_id) VALUES ?',
            [roleValues]
          );
        }
      }

      logger.info(`更新用户成功: ID=${id}, 操作人: ${req.user.username}`);

      res.json({
        success: true,
        message: '更新用户成功'
      });
    } catch (error) {
      logger.error('更新用户失败:', error);
      res.status(500).json({
        success: false,
        message: '更新用户失败'
      });
    }
  }

  async deleteUser(req, res) {
    try {
      const { id } = req.params;

      if (id === req.user.id) {
        return res.status(400).json({
          success: false,
          message: '不能删除当前登录用户'
        });
      }

      const [users] = await pool.query('SELECT id FROM users WHERE id = ?', [id]);

      if (users.length === 0) {
        return res.status(404).json({
          success: false,
          message: '用户不存在'
        });
      }

      await pool.query('DELETE FROM users WHERE id = ?', [id]);

      logger.info(`删除用户成功: ID=${id}, 操作人: ${req.user.username}`);

      res.json({
        success: true,
        message: '删除用户成功'
      });
    } catch (error) {
      logger.error('删除用户失败:', error);
      res.status(500).json({
        success: false,
        message: '删除用户失败'
      });
    }
  }

  async updateUserStatus(req, res) {
    try {
      const { id } = req.params;
      const { status } = req.body;

      if (id === req.user.id) {
        return res.status(400).json({
          success: false,
          message: '不能修改当前登录用户状态'
        });
      }

      const [users] = await pool.query('SELECT id FROM users WHERE id = ?', [id]);

      if (users.length === 0) {
        return res.status(404).json({
          success: false,
          message: '用户不存在'
        });
      }

      await pool.query(
        'UPDATE users SET status = ?, updated_by = ? WHERE id = ?',
        [status, req.user.id, id]
      );

      logger.info(`更新用户状态成功: ID=${id}, status=${status}, 操作人: ${req.user.username}`);

      res.json({
        success: true,
        message: '更新用户状态成功'
      });
    } catch (error) {
      logger.error('更新用户状态失败:', error);
      res.status(500).json({
        success: false,
        message: '更新用户状态失败'
      });
    }
  }

  async resetPassword(req, res) {
    try {
      const { id } = req.params;
      const { newPassword } = req.body;

      const [users] = await pool.query('SELECT id FROM users WHERE id = ?', [id]);

      if (users.length === 0) {
        return res.status(404).json({
          success: false,
          message: '用户不存在'
        });
      }

      const hashedPassword = await hashPassword(newPassword);

      await pool.query(
        'UPDATE users SET password = ?, updated_by = ? WHERE id = ?',
        [hashedPassword, req.user.id, id]
      );

      logger.info(`重置用户密码成功: ID=${id}, 操作人: ${req.user.username}`);

      res.json({
        success: true,
        message: '重置密码成功'
      });
    } catch (error) {
      logger.error('重置用户密码失败:', error);
      res.status(500).json({
        success: false,
        message: '重置密码失败'
      });
    }
  }
}

module.exports = new UserController();

File: code/Api/controllers/userController.go
package controllers

import (
	"net/http"
	"superhoneypotguard/database"
	"superhoneypotguard/middleware"
	"superhoneypotguard/models"
	"superhoneypotguard/utils"
	"time"

	"github.com/gin-gonic/gin"
)

type UserController struct{}

func NewUserController() *UserController {
	return &UserController{}
}

func (ctrl *UserController) GetList(c *gin.Context) {
	page := c.DefaultQuery("page", "1")
	pageSize := c.DefaultQuery("pageSize", "10")
	username := c.Query("username")
	status := c.Query("status")

	query := database.DB.Model(&models.User{})

	if username != "" {
		query = query.Where("username LIKE ?", "%"+username+"%")
	}

	if status != "" {
		query = query.Where("status = ?", status)
	}

	var total int64
	query.Count(&total)

	offset := (parseInt(page) - 1) * parseInt(pageSize)

	var users []models.User
	query.Offset(offset).Limit(parseInt(pageSize)).Order("created_at DESC").Find(&users)

	for i := range users {
		var roles []models.Role
		database.DB.Raw(`
			SELECT r.id, r.role_name, r.role_code
			FROM roles r
			INNER JOIN user_roles ur ON r.id = ur.role_id
			WHERE ur.user_id = ?
		`, users[i].ID).Scan(&roles)
		users[i].Roles = roles
	}

	utils.SuccessResponse(c, models.PaginatedResponse{
		List:     users,
		Total:    total,
		Page:     parseInt(page),
		PageSize: parseInt(pageSize),
	})
}

func (ctrl *UserController) GetById(c *gin.Context) {
	id := c.Param("id")

	var user models.User
	if err := database.DB.Where("id = ?", id).First(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "用户不存在")
		return
	}

	var roles []models.Role
	database.DB.Raw(`
		SELECT r.id, r.role_name, r.role_code
		FROM roles r
		INNER JOIN user_roles ur ON r.id = ur.role_id
		WHERE ur.user_id = ?
	`, user.ID).Scan(&roles)
	user.Roles = roles

	utils.SuccessResponse(c, user)
}

func (ctrl *UserController) Create(c *gin.Context) {
	var req models.CreateUserRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "参数验证失败")
		return
	}

	var count int64
	database.DB.Model(&models.User{}).Where("username = ? OR email = ?", req.Username, req.Email).Count(&count)
	if count > 0 {
		utils.ErrorResponse(c, http.StatusBadRequest, "用户名或邮箱已存在")
		return
	}

	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "密码加密失败")
		return
	}

	currentUser := middleware.GetCurrentUser(c)
	status := 1
	if req.Status != nil {
		status = *req.Status
	}

	user := models.User{
		Username:  req.Username,
		Password:  hashedPassword,
		Email:     req.Email,
		Phone:     req.Phone,
		RealName:  req.RealName,
		Status:    status,
		CreatedBy: &currentUser.UserID,
	}

	if err := database.DB.Create(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "创建用户失败")
		return
	}

	if len(req.RoleIDs) > 0 {
		for _, roleId := range req.RoleIDs {
			database.DB.Create(&models.UserRole{
				UserID:    user.ID,
				RoleID:    roleId,
				CreatedBy: &currentUser.UserID,
			})
		}
	}

	utils.SuccessResponse(c, gin.H{
		"id":       user.ID,
		"username": user.Username,
	})
}

func (ctrl *UserController) Update(c *gin.Context) {
	id := c.Param("id")

	var user models.User
	if err := database.DB.Where("id = ?", id).First(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "用户不存在")
		return
	}

	var req models.UpdateUserRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "参数验证失败")
		return
	}

	currentUser := middleware.GetCurrentUser(c)
	updates := map[string]interface{}{
		"email":     req.Email,
		"phone":     req.Phone,
		"real_name": req.RealName,
		"updated_by": currentUser.UserID,
	}

	if req.Status != nil {
		updates["status"] = *req.Status
	}

	if err := database.DB.Model(&user).Updates(updates).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "更新用户失败")
		return
	}

	if req.RoleIDs != nil {
		database.DB.Where("user_id = ?", user.ID).Delete(&models.UserRole{})

		for _, roleId := range *req.RoleIDs {
			database.DB.Create(&models.UserRole{
				UserID:    user.ID,
				RoleID:    roleId,
				CreatedBy: &currentUser.UserID,
			})
		}
	}

	utils.SuccessResponse(c, nil)
}

func (ctrl *UserController) Delete(c *gin.Context) {
	id := c.Param("id")
	currentUser := middleware.GetCurrentUser(c)

	if parseInt(id) == currentUser.UserID {
		utils.ErrorResponse(c, http.StatusBadRequest, "不能删除当前登录用户")
		return
	}

	var user models.User
	if err := database.DB.Where("id = ?", id).First(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "用户不存在")
		return
	}

	if err := database.DB.Delete(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "删除用户失败")
		return
	}

	utils.SuccessResponse(c, nil)
}

func (ctrl *UserController) UpdateStatus(c *gin.Context) {
	id := c.Param("id")
	currentUser := middleware.GetCurrentUser(c)

	if parseInt(id) == currentUser.UserID {
		utils.ErrorResponse(c, http.StatusBadRequest, "不能修改当前登录用户状态")
		return
	}

	var req models.UpdateUserStatusRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "参数验证失败")
		return
	}

	var user models.User
	if err := database.DB.Where("id = ?", id).First(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "用户不存在")
		return
	}

	if err := database.DB.Model(&user).Updates(map[string]interface{}{
		"status":     req.Status,
		"updated_by": currentUser.UserID,
	}).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "更新用户状态失败")
		return
	}

	utils.SuccessResponse(c, nil)
}

func (ctrl *UserController) ResetPassword(c *gin.Context) {
	id := c.Param("id")
	currentUser := middleware.GetCurrentUser(c)

	var req models.ResetPasswordRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "参数验证失败")
		return
	}

	var user models.User
	if err := database.DB.Where("id = ?", id).First(&user).Error; err != nil {
		utils.ErrorResponse(c, http.StatusNotFound, "用户不存在")
		return
	}

	hashedPassword, err := utils.HashPassword(req.NewPassword)
	if err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "密码加密失败")
		return
	}

	if err := database.DB.Model(&user).Updates(map[string]interface{}{
		"password":   hashedPassword,
		"updated_by": currentUser.UserID,
	}).Error; err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "重置密码失败")
		return
	}

	utils.SuccessResponse(c, nil)
}

func parseInt(s string) int {
	var result int
	for _, c := range s {
		if c >= '0' && c <= '9' {
			result = result*10 + int(c-'0')
		}
	}
	return result
}
